### PR TITLE
Add support for string roles & roles in reads

### PR DESCRIPTION
**Pull request by konne88**
This addresses bugs:
https://github.com/p4lang/p4runtime/issues/344
https://github.com/p4lang/p4runtime/issues/345

### CODE REVIEW

**Comment by konne88** [comment]
Antonin, I just got the tests to pass. Can you review. Thanks!

File: docs/v1/P4Runtime-Spec.mdk [hunk 1]
```diff
@@ -5168,7 +5167,7 @@ message SetForwardingPipelineConfigRequest {
     RECONCILE_AND_COMMIT = 5;
   }
   uint64 device_id = 1;
-  uint64 role_id = 2;
+  uint64 role = 2;
```
Comment (stolsma) [inline docs/v1/P4Runtime-Spec.mdk:5172]: Again doesn't compare to pb file, should be 'string role = 6' I think?

File: docs/v1/P4Runtime-Spec.mdk [hunk 2]
```diff
@@ -4728,7 +4727,7 @@ defined as follows:
 ~ Begin Proto
 message WriteRequest {
   uint64 device_id = 1;
-  uint64 role_id = 2;
+  uint64 role = 2;
```
[the remark this answers is not in the document]
Comment (konne88) [reply docs/v1/P4Runtime-Spec.mdk:4732]: Oh yeah, sorry I totally forgot that.

File: docs/v1/P4Runtime-Spec.mdk (hunk 1 above)
Comment (konne88) [reply docs/v1/P4Runtime-Spec.mdk]: Thanks for catching this!

File: docs/v1/P4Runtime-Spec.mdk [hunk 3]
```diff
@@ -4729,7 +4729,7 @@ defined as follows:
 ~ Begin Proto
 message WriteRequest {
   uint64 device_id = 1;
-  uint64 role = 2;
+  string role = 2;
```
Comment (stolsma) [inline docs/v1/P4Runtime-Spec.mdk:4732]: The number two needs to be 6 too...😉

File: docs/v1/P4Runtime-Spec.mdk [hunk 4]
```diff
@@ -5169,7 +5169,7 @@ message SetForwardingPipelineConfigRequest {
     RECONCILE_AND_COMMIT = 5;
   }
   uint64 device_id = 1;
-  uint64 role = 2;
+  string role = 2;
```
Comment (stolsma) [inline docs/v1/P4Runtime-Spec.mdk:5172]: Same here! 2 needs to be 6...😎

File: docs/v1/P4Runtime-Spec.mdk (hunk 3 above)
Comment (konne88) [reply docs/v1/P4Runtime-Spec.mdk:4732]: It's a good question of whether we want to keep the protos in the spec completely in sync with the real protos. As a middle ground for now, let me update the ids, but not add the deprecated fields in the spec (even though they are in the proto). I don't feel strongly about this though, happy to add the deprecated fields as well.

File: docs/v1/P4Runtime-Spec.mdk [hunk 5]
```diff
@@ -767,10 +767,9 @@ from making changes or even reading the state of the server.
 ## Default Role
 
 A controller can omit the role message in `MasterArbitrationUpdate`. This
-implies the "default role", which corresponds to "full pipeline access". This
-also implies that a default role has a `role.id` of 0 (default). If using a
-default role, all RPCs from the controller (&eg; `Write`) must set the `role_id`
-to 0.
+implies the "default role", which corresponds to "full pipeline access".
+If using a default role, all RPCs from the controller (&eg; `Write`) must
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk:771]: Should we mention explicitly that the empty string is therefore reserved for the default role and that it is not a valid name for a user-defined role? I noticed that you remove the sentence about role id 0.

File: docs/v1/P4Runtime-Spec.mdk [hunk 6]
```diff
@@ -4728,7 +4729,7 @@ defined as follows:
 ~ Begin Proto
 message WriteRequest {
   uint64 device_id = 1;
-  uint64 role_id = 2;
+  string role = 6;
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk:4732]: I am fine with the current solution of not including deprecated fields here. it is consistent with the fact that we do not describe deprecated fields such as `role_id` in the spec.

**Review by antoninbas — COMMENTED** [review]
could you squash the commits and include the following in the commit message body:
```
Fixes #344 
Fixes #345 
```
It will link the PR to these issues properly and close them automatically when the PR is merged. Thanks.

File: docs/v1/P4Runtime-Spec.mdk (hunk 5 above)
Comment (konne88) [reply docs/v1/P4Runtime-Spec.mdk:771]: I'm pretty sure that even in protobuf v3, empty string "" is different from unset string. Therefore, I think "" need not be a special case, and can just be like any other role.

**Comment by konne88** [comment]
I squashed the commits.

File: docs/v1/P4Runtime-Spec.mdk (hunk 5 above)
Comment (antoninbas) [reply docs/v1/P4Runtime-Spec.mdk:771]: No, string is a scalar field and empty string is the default
https://developers.google.com/protocol-buffers/docs/proto3#default

**Comment by antoninbas** [comment]
@konne-google you have merged conflicts, as I had to merge a separate PR

File: docs/v1/P4Runtime-Spec.mdk (hunk 5 above)
Comment (konne88) [reply docs/v1/P4Runtime-Spec.mdk:771]: Oh you are right, thanks! I though strings were different.

**Review by antoninbas — APPROVED** [review]
LGTM

**Comment by konne88** [comment]
Antonin, I don't think I can merge the request (The base branch restricts merging to authorized users.), could you do it for me? There is still a failing test, but I don't think I can fix it, it says:

docker: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit.

**Comment by antoninbas** [comment]
@konne88 I restarted the job, I'll merge when it passed

**Pull request merged**